### PR TITLE
Major rewrite of V6 Cryptography

### PR DESCRIPTION
**Pull request by danielcuthbert**
With V5 nearly ready, it was time we addressed the cryptography section so that it was more in-line with modern cryptographic approaches, including usability, security and addresses future concerns pertaining to Post-Quantum Cryptography. 

With the new section, we've added a number of sections, namely:

V6.1 Data Classification and Cryptographic Inventory
V6.2 General Requirements for Cryptographic Algorithms
V6.3 Cipher Algorithms
V6.3a Hashing and Hash-based Functions
V6.4 Random Values
V6.5 Secret Management
V6.6 Key Exchange Mechanisms
V6.7 In-Use Data Cryptography
V6.8 Post-Quantum Cryptography (PQC)

The main objective of this rewrite to help all on the journey to solid cryptography management. Those implementing cryptography in their apps MUST use strong cryptography; understand the techniques used, and maintain an inventory of their use cases, algorithms and keys, including obsolete assets, to address and mitigate risks derived from weak cryptography.

The terms used are taken from [NIST SP 800-57 Part 1](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf).

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119).

### CODE REVIEW

File: 5.0/en/0x14-V6-Cryptography.md [hunk 1]
```diff
@@ -2,62 +2,318 @@
...
+|     #     | Description                                                                                                                                                                                                                                                                                         | L1  | L2  | L3  | CWE |
+| :-------: | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :-: | :-: | :-: | :-: |
+| **6.1.1** | Verify that regulated data, such as private, health, and financial, is stored encrypted while at rest with at least 112-bit encryption, ideally 128-bit, such as Personally Identifiable Information (PII), sensitive personal information, or data assessed likely to be subject to the EU's GDPR. |     |  ✓  |  ✓  | 326 |
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:19]: "such as" is used twice in this sentence but is used for the same thing. It should be read as:

> Verify that regulated data, such as private, health, financial,  Personally Identifiable Information (PII), sensitive personal information, or data assessed likely to be subject to the EU's GDPR, is stored encrypted while at rest with at least 112-bit encryption, ideally 128-bit.

but does not parse like so for me in the current formulation.

Comment (randomstuff) [reply 5.0/en/0x14-V6-Cryptography.md:19]: Is it clear for everyone what "encrypted while at rest means"? For example, for databases should I encrypt the data at the column level (by the application itself) ? or is filesystem encryption enough ?

File: 5.0/en/0x14-V6-Cryptography.md [hunk 2]
```diff
@@ -2,62 +2,318 @@
...
+| **6.2.5** | Verify that insecure ciphers, including Triple-DES and Blowfish, are not used but secure ciphers and modes** such as AES with GCM are.                                                                                |  ✓  |  ✓  |  ✓  | 326 |
+| **6.2.6** | Verify that nonces, initialization vectors, and other single-use numbers are not used for more than one encryption key/data-element pair. The method of generation must be appropriate for the algorithm being used. |     |     |  ✓  | 326 |
+| **6.2.7** | Verify that encrypted data is authenticated via signatures, authenticated cipher modes, or HMAC to ensure that ciphertext is not altered by an unauthorized party.                                                   |     |     |  ✓  | 326 |
+| **6.2.8** | Verify that any authenticated signatures are operating in encrypt-then-MAC or encrypt-then-hash modes as required. party.                                                   |     |     |  ✓  | 326 |
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:61]: Spurious "party" here.

File: 5.0/en/0x14-V6-Cryptography.md [hunk 3]
```diff
@@ -2,62 +2,318 @@
...
+Implementations MUST choose from the ciphers in this list, in order of preference:
+
+| Symmetric Key Algorithms | Reference | L1 | L2 | L3 |
+|--|--|--|--|--|
+| AES-256 | [FIPS 197](https://csrc.nist.gov/pubs/fips/197/final) |     |  ✓  |  ✓  |
+| ChaCha20-Poly1305 | [RFC 8439](https://www.rfc-editor.org/info/rfc8439) |     |  ✓  |  ✓  |
+| AES-192 | [FIPS 197](https://csrc.nist.gov/pubs/fips/197/final) |     |  ✓  |  ✓  |
+| AES-128 | [FIPS 197](https://csrc.nist.gov/pubs/fips/197/final) |  ✓  |  ✓  |  ✓  |
+
+Any other cipher options MUST NOT be used.
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:74]: I guess some countries may requires the usage of other ciphers in some contexts (GOST, ShamMi). Should there be some provision be made for such cases?

Comment (randomstuff) [reply 5.0/en/0x14-V6-Cryptography.md:74]: Should some provision be made for new ciphers?

Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:74]: AES-* is a cipher while ChaCha20-Poly1305 is an AE. Should this list mention ChaCha20 only here?

File: 5.0/en/0x14-V6-Cryptography.md [hunk 4]
```diff
@@ -2,62 +2,318 @@
...
+The following cipher modes MUST NOT be used for any use case:
+
+| Encryption Mode | 
+|--|
+| ECB |
+| CFB | 
+| OFB | 
+| CTR | 
+
+### Key Wrapping
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:139]: Would there be need to explain for the layman when key wrapping should be used?

File: 5.0/en/0x14-V6-Cryptography.md [hunk 5]
```diff
@@ -2,62 +2,318 @@
...
-| **6.3.2** | [MODIFIED] Verify that GUIDs are created with an implementation of the GUID v4 algorithm which utilizes a cryptographically-secure pseudo-random number generator (CSPRNG). GUIDs created using other algorithm versions or using insufficiently secure pseudo-random number generators may be predictable. | | ✓ | ✓ | 338 |
-| **6.3.3** | Verify that random numbers are created with proper entropy even when the application is under heavy load, or that the application degrades gracefully in such circumstances. | | | ✓ | 338 |
+|     #     | Description                                                                                                 | L1  | L2  | L3  | CWE |
+| :-------: | :---------------------------------------------------------------------------------------------------------- | :-: | :-: | :-: | :-: |
+| **6.3.1** | All random numbers and strings intended to be non-guessable must be generated using a CSPRNG.               |     |  ✓  |  ✓  | 338 |
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:218]: Why is this one not L1?

File: 5.0/en/0x14-V6-Cryptography.md [hunk 6]
```diff
@@ -2,62 +2,318 @@
...
+| **6.3.2** | GUIDs must be created using the GUID v4 algorithm with CSPRNG.                                              |     |  ✓  |  ✓  | 338 |
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:219]: This one feelds weird to me. Why is this in cryptography? We are talking about UUIDs, right? There are legitimate use cases for other UUIDs versions. I feel this should be clarified.

File: 5.0/en/0x14-V6-Cryptography.md [hunk 7]
```diff
@@ -2,62 +2,318 @@
...
+| **6.3.3** | Random number generation must work properly under heavy system load, or the system must degrade gracefully. |     |     |  ✓  | 338 |
+
+### Approved RNG Methods and Algorithms
+
+| Name | Version/Reference | Notes | L1 | L2 | L3 |
+|:-:|:-:|:-:|:-:|:-:|:-:|
+| `/dev/random` | Linux 4.8+ [(Oct 2016)](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=818e607b57c94ade9824dad63a96c2ea6b21baf3), also found in iOS, Android, and other Linux-based POSIX operating systems. Based on [RFC7539](https://datatracker.ietf.org/doc/html/rfc7539) | Utilizing ChaCha20 stream. Found in iOS [`SecRandomCopyBytes`](https://developer.apple.com/documentation/security/secrandomcopybytes(_:_:_:)?language=objc) and Android [`Secure Random`](https://developer.android.com/reference/java/security/SecureRandom) with the correct settings provided to each. | ✓ | ✓ | ✓ |
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:226]: Mention that `/dev/urandom` and the Linux `getrandom()` are suitable as well?

File: 5.0/en/0x14-V6-Cryptography.md [hunk 8]
```diff
@@ -2,62 +2,318 @@
...
+
+Specifically, IKEv1 MUST NOT be used in production.
+
+## V6.7 In-Use Data Cryptography
+
+Protecting data while it is being processed is paramount. Techniques such as full memory encryption, encryption of data in transit, and ensuring data is encrypted as quickly as possible after use is recommended.
+
+|     #     | Description                                                                                                                                                                    | L1  | L2  | L3  | CWE |
+| :-------: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :-: | :-: | :-: | :-: |
+| **6.6.1** | Verify that full memory encryption is in use that protects sensitive data while it is in use, preventing access by unauthorized users or processes.                            |     |  ✓  |  ✓  | 798 |
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:308]: This seems drastic to me. For example, you cannot require that from user-side applications.

File: 5.0/en/0x14-V6-Cryptography.md [hunk 9]
```diff
@@ -2,62 +2,318 @@
...
-* [OWASP Cheat Sheet: Cryptographic Storage](https://cheatsheetseries.owasp.org/cheatsheets/Cryptographic_Storage_Cheat_Sheet.html)
-* [FIPS 140-3](https://csrc.nist.gov/pubs/fips/140-3/final)
+|     #     | Description                                                                                                                                                                                                                                 | L1  | L2  | L3  | CWE |
+| :-------: | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | :-: | :-: | :-: | :-: |
+| **6.7.1** | Verify that Quantum-Safe Algorithms, or quantum-resistant algorithms, such as lattice-based, hash-based, code-based, or multivariate cryptographic schemes, as replacements for vulnerable classical algorithms like RSA and ECC, are used. |     |  ✓  |  ✓  | 798 |
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:317]: >  Verify that Quantum-Safe Algorithms, or quantum-resistant algorithms, [...] **are used**.

I don't understand this. Is is saying that all the stuff above cannot be used?

File: 5.0/en/0x14-V6-Cryptography.md [hunk 10]
```diff
@@ -2,62 +2,318 @@
...
+| **6.7.2** | Verify that cryptographic systems are designed to allow for seamless upgrades to post-quantum cryptography, enabling the transition once PQC standards are fully established.                                                               |     |  ✓  |  ✓  | 798 |
+| **6.7.3** | Regularly monitor advancements in the field of post-quantum cryptography and align with emerging industry standards to remain prepared for quantum threats.                                                                                 |     |  ✓  |  ✓  | 798 |
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:319]: > Regularly monitor advancements in the field of post-quantum cryptography and align with emerging industry standards to remain prepared for quantum threats.  

You could say the same for cryptography in general :)

> Regularly monitor advancements in the field of cryptography and align with emerging industry standards to remain prepared for cryptographic threats.

File: 5.0/en/0x14-V6-Cryptography.md [hunk 11]
```diff
@@ -2,62 +2,318 @@
...
+| :-------: | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :-: | :-: | :-: | :-: |
+| **6.5.1** | Verify that industry-proven cryptographic algorithms, such as Diffie-Hellman groups, with a focus on ensuring that key exchange mechanisms use secure parameters to prevent man-in-the-middle attacks or cryptographic breaks, are used for key exchanges to prevent attacks on the key establishment process. |     |  ✓  |  ✓  | 798 |
+
+### Approved KEX Schemes
+
+A security strength of 112 bits or above MUST be ensured for all Key Exchange schemes, and their implementation MUST follow the parameter choices in the next table.
+
+| Scheme | Domain Parameters |
+|--|--|
+| RSA | k >= 2048 |
```
Comment (randomstuff) [inline 5.0/en/0x14-V6-Cryptography.md:267]: This lacks forward secrecy, right. Should this be approved?

**Comment by randomstuff** [comment]
Should there be some mention of forward secrecy in the requirements?

**Comment by elarlang** [comment]
The status here:
* there will be no easy merge for the PR to be done here, as the developed content has not followed our guidelines for keeping current numbering and using tags for tracking changes
* it does not contain changes that have already been made to V6

So the new content is here, but we need to figure out, how to get it into the document.

**Comment by danielcuthbert** [comment]
@elarlang 100% agree but V6 in the 4.x branch is stale and sadly isn't really useful anymore, hence this is that much of a breaking change. Totally accept this will be a mammoth amount of effort but that's why we have the summit this week and also myself and others getting it in shape

File: 5.0/en/0x14-V6-Cryptography.md (hunk 4 above)
Comment (danielcuthbert) [reply 5.0/en/0x14-V6-Cryptography.md:139]: 100% agree with this, yes there is indeed a need and we are adding it

**Review by jmanico — APPROVED** [review]

**Comment by elarlang** [comment]
This PR got in via #2234.

**Pull request closed without merging**